python-package/lightgbm/engine.py

```diff
@@ -4,21 +4,42 @@
 import copy
 from operator import attrgetter
 from pathlib import Path
+from typing import Any, Callable, Dict, List, Optional, Tuple, Union
 
 import numpy as np
 
 from . import callback
 from .basic import Booster, Dataset, LightGBMError, _ConfigAliases, _InnerPredictor, _log_warning
 from .compat import SKLEARN_INSTALLED, _LGBMGroupKFold, _LGBMStratifiedKFold
 
-
-def train(params, train_set, num_boost_round=100,
-          valid_sets=None, valid_names=None,
-          fobj=None, feval=None, init_model=None,
-          feature_name='auto', categorical_feature='auto',
-          early_stopping_rounds=None, evals_result=None,
-          verbose_eval=True, learning_rates=None,
-          keep_training_booster=False, callbacks=None):
+_LGBM_CustomObjectiveFunction = Callable[
+    [Union[List, np.ndarray], Dataset],
+    Tuple[Union[List, np.ndarray], Union[List, np.ndarray]]
+]
+_LGBM_CustomMetricFunction = Callable[
+    [Union[List, np.ndarray], Dataset],
+    Tuple[str, float, bool]
+]
+
+
+def train(
+    params: Dict[str, Any],
+    train_set: Dataset,
+    num_boost_round: int = 100,
+    valid_sets: Optional[List[Dataset]] = None,
+    valid_names: Optional[List[str]] = None,
+    fobj: Optional[_LGBM_CustomObjectiveFunction] = None,
+    feval: Optional[_LGBM_CustomMetricFunction] = None,
+    init_model: Optional[Union[str, Path, Booster]] = None,
+    feature_name: Union[List[str], str] = 'auto',
+    categorical_feature: Union[List[str], str] = 'auto',
+    early_stopping_rounds: Optional[int] = None,
+    evals_result: Optional[Dict[str, Any]] = None,
```

Comment:
This dict should be empty when passing to the function. Can we reflect this fact with typing?
```suggestion
    evals_result: Optional[Dict] = None,
```

Reply:
Oh interesting, I thought it could be anything but I see now that that dictionary is going to be emptied.

https://github.com/microsoft/LightGBM/blob/11d7608f2d755899a88bfa1c0fbc779b9e9b3505/python-package/lightgbm/callback.py#L90

Yes, there is a way to say in typing that something has a specific value, via the `Literal` type: https://www.python.org/dev/peps/pep-0586.

However, that type was only added to the standard library in Python 3.8, so I don't think it should be used in `lightgbm`.

```text
docker run -t python:3.7-slim python -c "from typing import Literal"
```

> Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: cannot import name 'Literal' from 'typing' (/usr/local/lib/python3.7/typing.py)

I think a better solution would be to leave the hint as `Dict[str, Any]`, and add to the docstring that anything in that dictionary is going to be removed by `lightgbm`.

That's still a correct documentation of what the eventual value will be. Using an inner type in containers like `Dict[]`, `List[]`, etc. doesn't necessarily imply that they'll be initialized with any contents.

e.g. this pattern is valid and common:

```python
def is_in_list(number: int, numbers: List[int]) -> bool:
    return number in numbers
```

even though that function might be called with an empty list, like `is_in_list(5, [])`

Reply:
I've proposed that docs update in https://github.com/microsoft/LightGBM/pull/4544/commits/2120d7135ba67e860c5b31b0be3498b8816bd2e4

Reply:
Perfect, thanks!